MQTT client attemps to reconnect indefinitely

I have tried to connect two MQTT clients with the same ID to the same broker without knowing that the clientID must be unique. In this situation both clients tries indefinitely to connect to the broker (normal behaviour) but when you stop one of them, the other one goes on attemping to reconnect with the debug message "MqttPahoMessageDrivenChannelAdapter : Attempting reconnect" forever. The only way to solve the problem is restart the application.

I don't know if this is a bug or the expected behaviour. These are the logs for the error of the instance that remains attemping to reconnect.

```
2017-02-02 14:50:39.487 DEBUG 27039 --- [nio-8081-exec-2] o.s.b.w.f.OrderedRequestContextFilter    : Cleared thread-bound request context: org.apache.catalina.connector.RequestFacade@696f298c
2017-02-02 14:50:45.384 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:45.384 DEBUG 27039 --- [mtfsParserInput] o.s.b.f.s.DefaultListableBeanFactory     : Returning cached instance of singleton bean 'taskScheduler'
2017-02-02 14:50:45.385 DEBUG 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:45.432 DEBUG 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Connected and subscribed to [mtfs-connector/tap]
2017-02-02 14:50:45.450 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:45.450 DEBUG 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:45.565 ERROR 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Error connecting or subscribing to [mtfs-connector/tap]

org.eclipse.paho.client.mqttv3.MqttException: Connection lost
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:146)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.EOFException: null
	at java.io.DataInputStream.readByte(DataInputStream.java:267)
	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:65)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:107)
	... 1 common frames omitted

2017-02-02 14:50:45.566 ERROR 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Exception while connecting and subscribing

org.eclipse.paho.client.mqttv3.MqttException: Client is currently disconnecting
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:31)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms.disconnect(ClientComms.java:455)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:632)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:601)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:608)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.connectAndSubscribe(MqttPahoMessageDrivenChannelAdapter.java:246)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.access$300(MqttPahoMessageDrivenChannelAdapter.java:47)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter$1.run(MqttPahoMessageDrivenChannelAdapter.java:287)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

2017-02-02 14:50:45.568 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:45.569 DEBUG 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:45.700 ERROR 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Error connecting or subscribing to [mtfs-connector/tap]

org.eclipse.paho.client.mqttv3.MqttException: Connection lost
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:146)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.EOFException: null
	at java.io.DataInputStream.readByte(DataInputStream.java:267)
	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:65)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:107)
	... 1 common frames omitted

2017-02-02 14:50:45.701 ERROR 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Exception while connecting and subscribing

org.eclipse.paho.client.mqttv3.MqttException: Client is disconnected
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:31)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms.disconnect(ClientComms.java:451)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:632)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:601)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:608)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.connectAndSubscribe(MqttPahoMessageDrivenChannelAdapter.java:246)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.access$300(MqttPahoMessageDrivenChannelAdapter.java:47)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter$1.run(MqttPahoMessageDrivenChannelAdapter.java:287)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

.
.
.

2017-02-02 14:50:46.462 DEBUG 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:46.548 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:46.549 DEBUG 27039 --- [ask-scheduler-8] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:46.549 ERROR 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Error connecting or subscribing to [mtfs-connector/tap]

org.eclipse.paho.client.mqttv3.MqttException: Connection lost
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:146)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.EOFException: null
	at java.io.DataInputStream.readByte(DataInputStream.java:267)
	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:65)
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:107)
	... 1 common frames omitted

2017-02-02 14:50:46.549 ERROR 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Exception while connecting and subscribing

org.eclipse.paho.client.mqttv3.MqttException: Client is disconnected
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:31)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms.disconnect(ClientComms.java:451)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:632)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:601)
	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.disconnect(MqttAsyncClient.java:608)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.connectAndSubscribe(MqttPahoMessageDrivenChannelAdapter.java:246)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter.access$300(MqttPahoMessageDrivenChannelAdapter.java:47)
	at org.springframework.integration.mqtt.inbound.MqttPahoMessageDrivenChannelAdapter$1.run(MqttPahoMessageDrivenChannelAdapter.java:287)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

2017-02-02 14:50:46.621 DEBUG 27039 --- [ask-scheduler-8] .m.i.MqttPahoMessageDrivenChannelAdapter : Connected and subscribed to [mtfs-connector/tap]
2017-02-02 14:50:46.637 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:46.639 DEBUG 27039 --- [ask-scheduler-5] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:46.699 DEBUG 27039 --- [ask-scheduler-5] .m.i.MqttPahoMessageDrivenChannelAdapter : Connected and subscribed to [mtfs-connector/tap]
2017-02-02 14:50:46.721 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:46.721 DEBUG 27039 --- [ask-scheduler-4] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:46.823 DEBUG 27039 --- [ask-scheduler-4] .m.i.MqttPahoMessageDrivenChannelAdapter : Connected and subscribed to [mtfs-connector/tap]
2017-02-02 14:50:46.824 ERROR 27039 --- [mtfsParserInput] .m.i.MqttPahoMessageDrivenChannelAdapter : Lost connection:Connection lost; retrying...
2017-02-02 14:50:46.825 DEBUG 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:46.909 DEBUG 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Connected and subscribed to [mtfs-connector/tap]
2017-02-02 14:50:49.490 DEBUG 27039 --- [nio-8081-exec-3] o.s.b.w.f.OrderedRequestContextFilter    : Bound request context to thread: org.apache.catalina.connector.RequestFacade@696f298c
2017-02-02 14:50:49.490 DEBUG 27039 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : DispatcherServlet with name 'dispatcherServlet' processing GET request for [/health]
2017-02-02 14:50:49.490 DEBUG 27039 --- [nio-8081-exec-3] o.s.b.a.e.mvc.EndpointHandlerMapping     : Looking up handler method for path /health
2017-02-02 14:50:49.491 DEBUG 27039 --- [nio-8081-exec-3] o.s.b.a.e.mvc.EndpointHandlerMapping     : Returning handler method [public java.lang.Object org.springframework.boot.actuate.endpoint.mvc.HealthMvcEndpoint.invoke(java.security.Principal)]
2017-02-02 14:50:49.491 DEBUG 27039 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : Last-Modified value for [/health] is: -1
2017-02-02 14:50:49.519 DEBUG 27039 --- [nio-8081-exec-3] m.m.a.RequestResponseBodyMethodProcessor : Written [UP {discoveryComposite=UP {discoveryClient=UP {services=[MtfsParser, consul]}}, diskSpace=UP {total=486303457280, free=411104645120, threshold=10485760}, refreshScope=UP {}, hystrix=UP {}, consul=UP {services={MtfsParser=[], consul=[]}, advertiseAddress=127.0.0.1, datacenter=mtfs-test, domain=consul., nodeName=os-lin-jra, bindAddress=0.0.0.0, clientAddress=127.0.0.1}}] as "application/json" using [org.springframework.http.converter.json.MappingJackson2HttpMessageConverter@7530090a]
2017-02-02 14:50:49.519 DEBUG 27039 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : Null ModelAndView returned to DispatcherServlet with name 'dispatcherServlet': assuming HandlerAdapter completed request handling
2017-02-02 14:50:49.519 DEBUG 27039 --- [nio-8081-exec-3] o.s.web.servlet.DispatcherServlet        : Successfully completed request
2017-02-02 14:50:49.520 DEBUG 27039 --- [nio-8081-exec-3] o.s.b.w.f.OrderedRequestContextFilter    : Cleared thread-bound request context: org.apache.catalina.connector.RequestFacade@696f298c
2017-02-02 14:50:55.566 DEBUG 27039 --- [ask-scheduler-6] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:55.701 DEBUG 27039 --- [ask-scheduler-7] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:55.933 DEBUG 27039 --- [ask-scheduler-3] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:56.016 DEBUG 27039 --- [ask-scheduler-8] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:56.115 DEBUG 27039 --- [ask-scheduler-5] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:56.206 DEBUG 27039 --- [ask-scheduler-4] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:56.342 DEBUG 27039 --- [ask-scheduler-2] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:56.462 DEBUG 27039 --- [sk-scheduler-10] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
2017-02-02 14:50:56.550 DEBUG 27039 --- [ask-scheduler-9] .m.i.MqttPahoMessageDrivenChannelAdapter : Attempting reconnect
```

--- a/spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java
+++ b/spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2002-2016 the original author or authors.
+ * Copyright 2002-2017 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,6 +17,7 @@
 package org.springframework.integration.mqtt.inbound;
 
 import java.util.Arrays;
+import java.util.Date;
 import java.util.concurrent.ScheduledFuture;
 
 import org.eclipse.paho.client.mqttv3.IMqttAsyncClient;
@@ -138,6 +139,7 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 
 	@Override
 	protected void doStart() {
+		Assert.state(getTaskScheduler() != null, "A 'taskScheduler' is required");
 		super.doStart();
 		try {
 			connectAndSubscribe();
@@ -149,7 +151,7 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 	}
 
 	@Override
-	protected void doStop() {
+	protected synchronized void doStop() {
 		cancelReconnect();
 		super.doStop();
 		if (this.client != null) {
@@ -219,7 +221,7 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 		}
 	}
 
-	private void connectAndSubscribe() throws MqttException {
+	private synchronized void connectAndSubscribe() throws MqttException {
 		MqttConnectOptions connectionOptions = this.clientFactory.getConnectionOptions();
 		this.cleanSession = connectionOptions.isCleanSession();
 		this.consumerStopAction = this.clientFactory.getConsumerStopAction();
@@ -259,10 +261,6 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 			if (this.applicationEventPublisher != null) {
 				this.applicationEventPublisher.publishEvent(new MqttSubscribedEvent(this, message));
 			}
-			// cancel() after the publish in case we are on that thread; a send to a QueueChannel would fail.
-			if (this.reconnectFuture != null) {
-				cancelReconnect();
-			}
 		}
 	}
 
@@ -275,19 +273,23 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 
 	private void scheduleReconnect() {
 		try {
-			this.reconnectFuture = this.getTaskScheduler().scheduleWithFixedDelay(() -> {
+			this.reconnectFuture = getTaskScheduler().schedule(() -> {
 				try {
 					if (logger.isDebugEnabled()) {
 						logger.debug("Attempting reconnect");
 					}
-					if (!MqttPahoMessageDrivenChannelAdapter.this.connected) {
-						connectAndSubscribe();
+					synchronized (MqttPahoMessageDrivenChannelAdapter.this) {
+						if (!MqttPahoMessageDrivenChannelAdapter.this.connected) {
+							connectAndSubscribe();
+							MqttPahoMessageDrivenChannelAdapter.this.reconnectFuture = null;
+						}
 					}
 				}
 				catch (MqttException e) {
 					logger.error("Exception while connecting and subscribing", e);
+					scheduleReconnect();
 				}
-			}, this.recoveryInterval);
+			}, new Date(System.currentTimeMillis() + this.recoveryInterval));
 		}
 		catch (Exception e) {
 			logger.error("Failed to schedule reconnect", e);
@@ -295,7 +297,7 @@ public class MqttPahoMessageDrivenChannelAdapter extends AbstractMqttMessageDriv
 	}
 
 	@Override
-	public void connectionLost(Throwable cause) {
+	public synchronized void connectionLost(Throwable cause) {
 		this.logger.error("Lost connection:" + cause.getMessage() + "; retrying...");
 		this.connected = false;
 		scheduleReconnect();

--- a/spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/MqttAdapterTests.java
+++ b/spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/MqttAdapterTests.java
@@ -17,6 +17,7 @@
 package org.springframework.integration.mqtt;
 
 import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertSame;
@@ -24,6 +25,8 @@ import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.BDDMockito.willAnswer;
 import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.doReturn;
 import static org.mockito.Mockito.mock;
@@ -39,11 +42,13 @@ import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
 import javax.net.SocketFactory;
 
 import org.aopalliance.intercept.MethodInterceptor;
+import org.apache.commons.logging.Log;
 import org.eclipse.paho.client.mqttv3.IMqttToken;
 import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
 import org.eclipse.paho.client.mqttv3.MqttCallback;
@@ -57,6 +62,7 @@ import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
 import org.junit.Test;
 
 import org.springframework.aop.framework.ProxyFactoryBean;
+import org.springframework.beans.DirectFieldAccessor;
 import org.springframework.beans.factory.BeanFactory;
 import org.springframework.context.ApplicationEventPublisher;
 import org.springframework.context.annotation.AnnotationConfigApplicationContext;
@@ -77,6 +83,7 @@ import org.springframework.integration.mqtt.support.DefaultPahoMessageConverter;
 import org.springframework.integration.test.util.TestUtils;
 import org.springframework.messaging.Message;
 import org.springframework.messaging.support.GenericMessage;
+import org.springframework.scheduling.TaskScheduler;
 import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
 
 /**
@@ -301,6 +308,7 @@ public class MqttAdapterTests {
 		}
 		assertThat(event, instanceOf(MqttSubscribedEvent.class));
 		assertEquals("Connected and subscribed to [baz, fix]", ((MqttSubscribedEvent) event).getMessage());
+		taskScheduler.destroy();
 	}
 
 	@Test
@@ -369,6 +377,34 @@ public class MqttAdapterTests {
 		ctx.close();
 	}
 
+	@Test
+	public void testReconnect() throws Exception {
+		final MqttAsyncClient client = mock(MqttAsyncClient.class);
+		MqttPahoMessageDrivenChannelAdapter adapter = buildAdapter(client, null, ConsumerStopAction.UNSUBSCRIBE_NEVER);
+		adapter.setRecoveryInterval(10);
+		Log logger = spy(TestUtils.getPropertyValue(adapter, "logger", Log.class));
+		new DirectFieldAccessor(adapter).setPropertyValue("logger", logger);
+		given(logger.isDebugEnabled()).willReturn(true);
+		final AtomicInteger attemptingReconnectCount = new AtomicInteger();
+		willAnswer(i -> {
+			if (attemptingReconnectCount.getAndIncrement() == 0) {
+				adapter.connectionLost(new RuntimeException("while schedule running"));
+			}
+			i.callRealMethod();
+			return null;
+		}).given(logger).debug("Attempting reconnect");
+		ThreadPoolTaskScheduler taskScheduler = new ThreadPoolTaskScheduler();
+		taskScheduler.initialize();
+		adapter.setTaskScheduler(taskScheduler);
+		adapter.start();
+		adapter.connectionLost(new RuntimeException("initial"));
+		Thread.sleep(1000);
+		// the following assertion should be equalTo, but leq to protect against a slow CI server
+		assertThat(attemptingReconnectCount.get(), lessThanOrEqualTo(2));
+		adapter.stop();
+		taskScheduler.destroy();
+	}
+
 	private MqttPahoMessageDrivenChannelAdapter buildAdapter(final MqttAsyncClient client, Boolean cleanSession,
 			ConsumerStopAction action) throws MqttException, MqttSecurityException {
 		DefaultMqttPahoClientFactory factory = new DefaultMqttPahoClientFactory() {
@@ -394,6 +430,7 @@ public class MqttAdapterTests {
 		MqttPahoMessageDrivenChannelAdapter adapter = new MqttPahoMessageDrivenChannelAdapter("client", factory, "foo");
 		adapter.setApplicationEventPublisher(mock(ApplicationEventPublisher.class));
 		adapter.setOutputChannel(new NullChannel());
+		adapter.setTaskScheduler(mock(TaskScheduler.class));
 		adapter.afterPropertiesSet();
 		return adapter;
 	}

